### PR TITLE
Scaler apis

### LINKED ISSUE #9182: Consume new Scaler APIs from function host
There was the change in webjobs SDK aiming to unify scaling logic to be reused in SC V3:
https://github.com/Azure/azure-webjobs-sdk/pull/2954
Bunch of scaling interfaces/classes were moved from function hhost to webjobs sdk, renamed or updated. Function host have to consume these new APIs from webjobs sdk to avoid duplication.

### PR DESCRIPTION
**Pull request by alrod**
### Issue describing the changes in this PR

resolves #9182

### Pull request checklist

* [x] My changes **do not** require documentation changes
    * [ ] Otherwise: Documentation issue linked to PR
* [x] My changes **should not** be added to the release notes for the next release
    * [ ] Otherwise: I've added my notes to `release_notes.md`
* [x] My changes **do not** need to be backported to a previous version
    * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [x] My changes **do not** require diagnostic events changes
    * Otherwise: I have added/updated all related diagnostic events and their documentation (Documentation issue linked to PR)
* [x] I have added all required tests (Unit tests, E2E tests)

### Additional information

Additional PR information

### CODE REVIEW

**Comment by liliankasem** [comment]
Please link a related issue and/or update PR description with details on changes made - thanks!

File: src/WebJobs.Script/ScriptHostBuilderExtensions.cs [hunk 1]
```diff
@@ -309,9 +310,15 @@ public static IHostBuilder AddScriptHostCore(this IHostBuilder builder, ScriptAp
                 services.AddOptions<ScaleOptions>()
                     .Configure<IConfiguration>((o, c) =>
                     {
-                        c.GetSection(ConfigurationSectionNames.JobHost)
-                         .GetSection(ConfigurationSectionNames.Scale)
-                         .Bind(o);
+                        var scaleSection = c.GetSection(ConfigurationSectionNames.JobHost).GetSection(ConfigurationSectionNames.Scale);
+                        if (scaleSection.Exists())
```
Comment (mathewc) [inline src/WebJobs.Script/ScriptHostBuilderExtensions.cs]: Rather than this new conditional logic here, I think you should leave this code as is, and register an options setup (e.g. see ConcurrencyOptionsSetup). That's the way this is usually done.

File: src/WebJobs.Script.WebHost/Controllers/HostController.cs [hunk 2]
```diff
@@ -318,7 +318,7 @@ public async Task<IActionResult> GetScaleStatus([FromBody] ScaleStatusContext co
 
             // TEMP: Once https://github.com/Azure/azure-functions-host/issues/5161 is fixed, we should take
             // FunctionsScaleManager as a parameter.
-            if (Utility.TryGetHostService(scriptHostManager, out FunctionsScaleManager scaleManager))
+            if (Utility.TryGetHostService(scriptHostManager, out IScaleManager scaleManager))
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Controllers/HostController.cs]: nit. Fix the comment above referencing FunctionsScaleManager

File: src/WebJobs.Script/ScriptHostBuilderExtensions.cs [hunk 3]
```diff
@@ -328,15 +335,15 @@ public static IHostBuilder AddScriptHostCore(this IHostBuilder builder, ScriptAp
...
-                    services.TryAddEnumerable(ServiceDescriptor.Singleton<IHostedService, FunctionsScaleMonitorService>());
-                }
-                services.TryAddSingleton<FunctionsScaleManager>();
-
                 services.AddSingleton<IHostOptionsProvider, HostOptionsProvider>();
             });
 
+            if (SystemEnvironment.Instance.IsRuntimeScaleMonitoringEnabled())
+            {
+                // This is needed only to add ScaleMonitorService, all other services for scale are added by the time of the call
```
Comment (mathewc) [inline src/WebJobs.Script/ScriptHostBuilderExtensions.cs]: If it's only ScaleMonitorService that we need to add, I wonder if we shouldn't just add it in all cases, and leverage the existing logic in the service that only starts the service when RTS is enabled?

Comment (alrod) [reply src/WebJobs.Script/ScriptHostBuilderExtensions.cs]: yes, the latest webjobs sdk has this.

File: src/WebJobs.Script/ScriptHostBuilderExtensions.cs (hunk 1 above)
Comment (alrod) [reply src/WebJobs.Script/ScriptHostBuilderExtensions.cs]: Created ScaleOptionsSetup.

File: src/WebJobs.Script.WebHost/Controllers/HostController.cs [hunk 4]
```diff
@@ -317,8 +317,8 @@ public async Task<IActionResult> GetScaleStatus([FromBody] ScaleStatusContext co
             }
 
             // TEMP: Once https://github.com/Azure/azure-functions-host/issues/5161 is fixed, we should take
-            // FunctionsScaleManager as a parameter.
-            if (Utility.TryGetHostService(scriptHostManager, out FunctionsScaleManager scaleManager))
+            // ScaleManager as a parameter.
+            if (Utility.TryGetHostService(scriptHostManager, out ScaleManager scaleManager))
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Controllers/HostController.cs]: This will become IScaleStatusProvider

File: src/WebJobs.Script/Config/ScaleOptionsSetup.cs [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+
+        public ScaleOptionsSetup(IConfiguration configuration, IEnvironment environment)
+        {
+            _configuration = configuration;
+            _environment = environment;
+        }
+
+        public void Configure(ScaleOptions options)
+        {
+            var jobHostSection = _configuration.GetSection(ConfigurationSectionNames.JobHost);
```
Comment (mathewc) [inline src/WebJobs.Script/Config/ScaleOptionsSetup.cs]: SDK is already binding these options. Do we need this? How does it interact with the other bind? Also, seems like code elsewhere in this PR is also duplicating this bind logic.

Comment (alrod) [reply src/WebJobs.Script/Config/ScaleOptionsSetup.cs]: SDK does not configure the options - we can pass the options into `ConfigureWebJobsScale`. `ConfigureWebJobsScale `is not called from the function host so we need a way to configure it the function host. I moved the code to setup class from [here](https://github.com/Azure/azure-functions-host/pull/9131/files/#diff-0520ea7869aafe4adf4e46cee524515169549321187cc8ecf473409fe070b53cL309).

File: src/WebJobs.Script.WebHost/Controllers/HostController.cs [hunk 6]
```diff
@@ -317,8 +317,8 @@ public async Task<IActionResult> GetScaleStatus([FromBody] ScaleStatusContext co
             }
 
             // TEMP: Once https://github.com/Azure/azure-functions-host/issues/5161 is fixed, we should take
-            // FunctionsScaleManager as a parameter.
-            if (Utility.TryGetHostService(scriptHostManager, out FunctionsScaleManager scaleManager))
+            // ScaleManager as a parameter.
+            if (Utility.TryGetHostService(scriptHostManager, out IScaleStatusProvider scaleManager))
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Controllers/HostController.cs]: naming/comments no longer match type

File: src/WebJobs.Script.WebHost/WebScriptHostBuilderExtension.cs [hunk 7]
```diff
@@ -131,7 +131,7 @@ public static class WebScriptHostBuilderExtension
                         // EasyAuth must go after CORS, as CORS preflight requests can happen before authentication
                         services.TryAddEnumerable(ServiceDescriptor.Singleton<IJobHostHttpMiddleware, JobHostEasyAuthMiddleware>());
                     }
-                    services.TryAddSingleton<IScaleMetricsRepository, TableStorageScaleMetricsRepository>();
+                    services.AddSingleton<IScaleMetricsRepository, TableStorageScaleMetricsRepository>();
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/WebScriptHostBuilderExtension.cs:134]: Why the change here?

File: src/WebJobs.Script/Config/ScaleOptionsSetup.cs (hunk 5 above)
Comment (mathewc) [reply src/WebJobs.Script/Config/ScaleOptionsSetup.cs]: I'm talking about the code [here](https://github.com/Azure/azure-webjobs-sdk/blob/55f3d9a7c8e0f38c9ad31b94585c9d395eeb7f57/src/Microsoft.Azure.WebJobs.Host/Hosting/WebJobsServiceCollectionExtensions.cs#L196) in the SDK that binds these same options. Base SDK binds already - why do we do it again?

File: src/WebJobs.Script/Config/ScaleOptionsSetup.cs [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+            var scaleSection = jobHostSection.GetSection(ConfigurationSectionNames.Scale);
+
+            if (scaleSection.Exists())
+            {
+                scaleSection.Bind(options);
+            }
+            else
+            {
+                options.IsRuntimeScalingEnabled = _environment.IsRuntimeScaleMonitoringEnabled();
```
Comment (mathewc) [inline src/WebJobs.Script/Config/ScaleOptionsSetup.cs]: I don't understand why this defaulting is conditional - don't you want this to run regardless of whether the configuration section exists?

File: src/WebJobs.Script/ScriptHostBuilderExtensions.cs [hunk 9]
```diff
@@ -102,6 +103,12 @@ public static IHostBuilder AddScriptHost(this IHostBuilder builder, Action<Scrip
                 {
                     configBuilder.Add(new HostJsonFileConfigurationSource(applicationOptions, SystemEnvironment.Instance, loggerFactory, metricsLogger));
                 }
+                // Adding hsoting config into job host configuration
```
Comment (mathewc) [inline src/WebJobs.Script/ScriptHostBuilderExtensions.cs]: spelling

File: src/WebJobs.Script/ScriptHostBuilderExtensions.cs [hunk 10]
```diff
@@ -102,6 +103,12 @@ public static IHostBuilder AddScriptHost(this IHostBuilder builder, Action<Scrip
...
+                IConfiguration scriptHostConfiguration = applicationOptions.RootServiceProvider.GetService<IConfiguration>();
+                if (scriptHostConfiguration != null)
+                {
+                    configBuilder.AddConfiguration(scriptHostConfiguration.GetSection(ScriptConstants.FunctionsHostingConfigSectionName));
```
Comment (mathewc) [inline src/WebJobs.Script/ScriptHostBuilderExtensions.cs:110]: What's going on here? There's already a FunctionsHostingConfigProvider registered. Why do we need changes to this for this PR?

File: test/WebJobs.Script.Tests/Controllers/Admin/HostControllerTests.cs [hunk 11]
```diff
@@ -350,5 +357,49 @@ public async Task ResumeHost_HostRunning_DrainModeNotEnabled_DoesNotStartNewHost
             Assert.Equal(expectedBody.State, (result.Value as ResumeStatus).State);
             scriptHostManagerMock.Verify(p => p.RestartHostAsync(It.IsAny<CancellationToken>()), Times.Never());
         }
+
+        //private ScaleManager CreateScaleManager()
```
Comment (mathewc) [inline test/WebJobs.Script.Tests/Controllers/Admin/HostControllerTests.cs]: clean this up

File: test/WebJobs.Script.Tests/Scale/ScaleMonitorServiceTests.cs [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+namespace Microsoft.Azure.WebJobs.Script.Tests.Scale
+{
+    public class ScaleMonitorServiceTests
+    {
+        [Theory]
+        [InlineData(true, true)]
+        [InlineData(false, true)]
+        [InlineData(true, false)]
+        [InlineData(false, false)]
+        public async Task ScaleMonitorService_RegistersExpected(bool runtimeScaleMonitoringEnabled, bool targetBaseScalingEnabled)
```
Comment (mathewc) [inline test/WebJobs.Script.Tests/Scale/ScaleMonitorServiceTests.cs]: What is this test testing? ScaleMonitorService doesn't register anything. Seems like you're just trying to test options defaulting - so shouldn't those go in an options setup test?

File: test/WebJobs.Script.Tests/Scale/ScaleMonitorServiceTests.cs [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+                        Assert.Contains(testLoggerProvider.GetAllLogMessages(), x => x.FormattedMessage.Contains("\"IsTargetScalingEnabled\": false"));
+                    }
+                }
+                else
+                {
+                    Assert.DoesNotContain(testLoggerProvider.GetAllLogMessages(), x => x.FormattedMessage.StartsWith("Scale monitor service is started."));
+                }
+
+                // Clean
+                Environment.SetEnvironmentVariable(EnvironmentSettingNames.FunctionsRuntimeScaleMonitoringEnabled, string.Empty);
```
Comment (mathewc) [inline test/WebJobs.Script.Tests/Scale/ScaleMonitorServiceTests.cs]: Generally for things like this you put in a finally block to ensure it runs regardless of whether exceptions occur

File: src/WebJobs.Script.WebHost/WebScriptHostBuilderExtension.cs (hunk 7 above)
Comment (alrod) [reply src/WebJobs.Script.WebHost/WebScriptHostBuilderExtension.cs:134]: We wat to replace default `NullMetricsRepository`

File: src/WebJobs.Script/Config/ScaleOptionsSetup.cs (hunk 5 above)
Comment (alrod) [reply src/WebJobs.Script/Config/ScaleOptionsSetup.cs]: You are right, I missed the binding. Removing function host binding..

**Comment by fabiocav** [comment]
> Please link a related issue and/or update PR description with details on changes made - thanks!

Can you please follow up on this comment from @liliankasem or provide more information in the PR description?

File: src/WebJobs.Script/ScriptHostBuilderExtensions.cs (hunk 10 above)
Comment (alrod) [reply src/WebJobs.Script/ScriptHostBuilderExtensions.cs:110]: `FunctionsHostingConfigProvider` is registered for `ScriptHost`. `JobHost` has differerent set of `IConfiguration`. We want to copy `FunctionsHostingConfig` to `JobHost` to have access for the features in webjobs sdk.

File: test/WebJobs.Script.Tests/Scale/ScaleMonitorServiceTests.cs (hunk 12 above)
Comment (alrod) [reply test/WebJobs.Script.Tests/Scale/ScaleMonitorServiceTests.cs]: yes, renamed

File: test/WebJobs.Script.Tests/Configuration/ScaleMonitorOptionsSetupTests.cs [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+    {
+        [Theory]
+        [InlineData(true, true)]
+        [InlineData(false, true)]
+        [InlineData(true, false)]
+        [InlineData(false, false)]
+        public void ScaleOptionsSetup_Works_As_Expected(bool runtimeScaleMonitoringEnabled, bool targetBaseScalingEnabled)
+        {
+            var testEnvironment = new TestEnvironment();
+            var testMetricLogger = new TestMetricsLogger();
```
Comment (pragnagopa) [inline test/WebJobs.Script.Tests/Configuration/ScaleMonitorOptionsSetupTests.cs]: remove unused variables - testMetricLogger, configurationBuilder , testProfileManager, configuration

File: test/WebJobs.Script.Tests/Configuration/ScaleMonitorOptionsSetupTests.cs [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+
+            ScaleOptionsSetup setup = new ScaleOptionsSetup(testEnvironment);
+            ScaleOptions options = new ScaleOptions();
+
+            setup.Configure(options);
+
+            if (runtimeScaleMonitoringEnabled)
+            {
+                Assert.Equal(runtimeScaleMonitoringEnabled, options.IsRuntimeScalingEnabled);
+                testEnvironment.SetEnvironmentVariable(EnvironmentSettingNames.FunctionsRuntimeScaleMonitoringEnabled, "1");
```
Comment (pragnagopa) [inline test/WebJobs.Script.Tests/Configuration/ScaleMonitorOptionsSetupTests.cs]: any reason to set this again?

File: test/WebJobs.Script.Tests/Scale/FunctionsScaleManagerTests.cs [hunk 16]
```diff
@@ -1,387 +0,0 @@
...
-            List<IScaleMonitor> monitors = new List<IScaleMonitor>
-            {
-                monitor1
-            };
-            if (!tbsEnabled)
-            {
-                monitors.Add(monitor2);
-                monitors.Add(monitor3);
-            }
-            _monitorManagerMock.Setup(p => p.GetMonitors()).Returns(monitors);
```
Comment (pragnagopa) [inline test/WebJobs.Script.Tests/Scale/FunctionsScaleManagerTests.cs:120]: Is there test coverage for these scenarios in WebJobs SDK?

**Review by pragnagopa — COMMENTED** [review]
added minor comments

File: src/WebJobs.Script.WebHost/Controllers/HostController.cs (hunk 6 above)
Comment (pragnagopa) [reply src/WebJobs.Script.WebHost/Controllers/HostController.cs]: nit: one more in Line 330: FunctionsScaleManager does not exist anymore

File: test/WebJobs.Script.Tests/Scale/FunctionsScaleManagerTests.cs (hunk 16 above)
Comment (alrod) [reply test/WebJobs.Script.Tests/Scale/FunctionsScaleManagerTests.cs:120]: yes: https://github.com/Azure/azure-webjobs-sdk/blob/dev/test/Microsoft.Azure.WebJobs.Host.UnitTests/Scale/ScaleManagerTests.cs

**Review by pragnagopa — APPROVED** [review]

File: src/WebJobs.Script.WebHost/Controllers/HostController.cs (hunk 6 above)
Comment (mathewc) [reply src/WebJobs.Script.WebHost/Controllers/HostController.cs]: Change the name of the parameter to match as well - "scaleStatusProvider"

File: src/WebJobs.Script.WebHost/Controllers/HostController.cs [hunk 17]
```diff
@@ -327,7 +327,7 @@ public async Task<IActionResult> GetScaleStatus([FromBody] ScaleStatusContext co
             {
                 // This case should never happen. Because this action is marked RequiresRunningHost,
                 // it's only invoked when the host is running, and if it's running, we'll have access
-                // to the FunctionsScaleManager.
+                // to the IScriptHostManager.
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Controllers/HostController.cs]: This should be IScaleStatusProvider

File: src/WebJobs.Script.WebHost/WebScriptHostBuilderExtension.cs (hunk 7 above)
Comment (mathewc) [reply src/WebJobs.Script.WebHost/WebScriptHostBuilderExtension.cs:134]: Ah - because WebJobs SDK didn't used to register an instance, we could use TryAdd. Now it registers a default instance so we have to override.

File: src/WebJobs.Script/ScriptHostBuilderExtensions.cs [hunk 18]
```diff
@@ -306,13 +313,7 @@ public static IHostBuilder AddScriptHostCore(this IHostBuilder builder, ScriptAp
...
                          .Bind(o);
                     });
-                services.AddOptions<ScaleOptions>()
-                    .Configure<IConfiguration>((o, c) =>
-                    {
-                        c.GetSection(ConfigurationSectionNames.JobHost)
-                         .GetSection(ConfigurationSectionNames.Scale)
-                         .Bind(o);
-                    });
+                services.ConfigureOptions<ScaleOptionsSetup>();
```
Comment (mathewc) [inline src/WebJobs.Script/ScriptHostBuilderExtensions.cs:316]: I think this looks right now. My main concern is/was whether we'd have any breaking changes with apps that might be using app settings to configure ScaleOptions now, if we're changing the config section names we're binding to. The SDK logic is [here](https://github.com/Azure/azure-webjobs-sdk/blob/474b704e2711977d5ea106f19bc060970b5c151e/src/Microsoft.Azure.WebJobs.Host/Hosting/WebJobsServiceCollectionExtensions.cs#L195) and it uses a helper that allows the Functions Host to override the section name which it does [here](https://github.com/Azure/azure-functions-host/blob/d207a4380e18aa79a7545762bd2f5107f8338893/src/WebJobs.Script/ScriptHostBuilderExtensions.cs#L396). So I think we're good, I just need you to be sure/confirm :)

File: test/WebJobs.Script.Tests/Configuration/ScaleMonitorOptionsSetupTests.cs [hunk 19]
```diff
@@ -0,0 +1,45 @@
...
+namespace Microsoft.Azure.WebJobs.Script.Tests.Configuration
+{
+    public class ScaleMonitorOptionsSetupTests
+    {
+        [Theory]
+        [InlineData(true, true)]
+        [InlineData(false, true)]
+        [InlineData(true, false)]
+        [InlineData(false, false)]
+        public void ScaleOptionsSetup_Works_As_Expected(bool runtimeScaleMonitoringEnabled, bool targetBaseScalingEnabled)
```
Comment (mathewc) [inline test/WebJobs.Script.Tests/Configuration/ScaleMonitorOptionsSetupTests.cs]: nit: ScaleOptionsSetup_ConfiguresExpectedDefaults

File: src/WebJobs.Script/ScriptHostBuilderExtensions.cs (hunk 18 above)
Comment (alrod) [reply src/WebJobs.Script/ScriptHostBuilderExtensions.cs:316]: yes, there is no breaking change - the section names remain the same.

**Review by mathewc — APPROVED** [review]
